src/psyclone/configuration.py

```diff
@@ -0,0 +1,393 @@
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors: R. W. Ford and A. R. Porter, STFC Daresbury Lab
+
+'''
+PSyclone configuration management module.
+
+Deals with reading the config file and storing default settings.
+'''
+
+from __future__ import absolute_import
+import os
+
+
+# Name of the config file we search for
+_FILE_NAME = "psyclone.cfg"
+
+
+class ConfigurationError(Exception):
+    '''
+    Class for all configuration-related errors.
+    '''
+    def __init__(self, value, config=None):
+        '''
+        :param str value: error message
+        :param config: the Config object associated with the error
+        :type :py:class:`psyclone.configuration.Config`
+        '''
+        Exception.__init__(self, value)
+        self.value = "PSyclone configuration error"
+        if config:
+            self.value += " (file={0})".format(config.filename)
+        self.value += ": "+value
+
+    def __str__(self):
+        return repr(self.value)
+
+
+def _str_to_list(svalue):
+    '''
+    Helper routine to take a string containing a list of values and return
+    a list of those values.
+
+    :param str svalue: string containing space- or comma-delimited list
+                       of items
+    :returns: list of strings
+    :rtype: list
+    '''
+    if "," in svalue:
+        # Comma delimited
+        return [
+            str(item.strip()) for item in svalue.split(",")
+            if item.strip() != '']
+    else:
+        # Space delimited
+        return [
+            str(item.strip()) for item in svalue.split(" ")
+            if item.strip() != '']
+
+
+class ConfigFactory(object):
+    '''
+    Creates a singleton instance of our Config object.
+    '''
+    _instance = None  # Our single Config object
+
+    def __init__(self, config_file=None):
+        '''
+        Create our singleton Config object. If config_file is specified
+        then we throw-away the old Config and create a new one.
+
+        :param str config_file: Specific configuration file to use when
+                                creating Config object
+        '''
+        if not ConfigFactory._instance or config_file:
+            # Create a Config object if we've not already got one or if the
+            # caller has specified a particular file
+            ConfigFactory._instance = Config(config_file)
+
+    @staticmethod
+    def create():
+        '''
+        :returns: the singleton Config instance
+        :rtype: :py:class:`psyclone.config.Config`
+        '''
+        return ConfigFactory._instance
+
+
+class Config(object):
```

Comment:
I have a slight concern that this class provides information for all API's irrespective of which API we have chosen. I understand there is a chicken and egg issue here but could the specific sections have their own classes, or would that make things worse? I am happy to stick with what we have at the moment as there are not many parameters but might there be problems in the future e.g. if two different api-specific sections have different defaults for the same thing?

Reply:
Yes, I agree. I've therefore changed the implementation so that Dynamo 0.3-specific settings are now stored in their own object (which itself belongs to the Config object). I've written some developer documentation about this too.